extensions/typescript-language-features/src/utils/logLevelMonitor.ts

```diff
@@ -0,0 +1,110 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from 'vscode';
+import { Disposable } from './dispose';
+import { localize } from '../tsServer/versionProvider';
+import { TsServerLogLevel } from './configuration';
+
+export class LogLevelMonitor extends Disposable {
+
+	private static readonly logLevelConfigKey = 'typescript.tsserver.log';
+	private static readonly logLevelChangedStorageKey = 'typescript.tsserver.logLevelChanged';
+	private static readonly doNotPromptLogLevelStorageKey = 'typescript.tsserver.doNotPromptLogLevel';
+
+	constructor(private readonly context: vscode.ExtensionContext) {
+		super();
+
+		this._register(vscode.workspace.onDidChangeConfiguration(this.onConfigurationChange, this, this._disposables));
+
+		if (this.shouldNotifyExtendedLogging()) {
+			this.notifyExtendedLogging();
+		}
+	}
+
+	private onConfigurationChange(event: vscode.ConfigurationChangeEvent) {
+		const logLevelChanged = event.affectsConfiguration(LogLevelMonitor.logLevelConfigKey);
+		if (!logLevelChanged) {
+			return;
+		}
+		this.context.globalState.update(LogLevelMonitor.logLevelChangedStorageKey, new Date());
+	}
+
+	private get logLevel(): TsServerLogLevel {
+		return TsServerLogLevel.fromString(vscode.workspace.getConfiguration().get<string>(LogLevelMonitor.logLevelConfigKey, 'off'));
+	}
+
+	/**
+	 * Last date change if it exists and can be parsed as a date,
+	 * otherwise undefined.
+	 */
+	private get lastLogLevelChange(): Date | undefined {
+		const lastChange = this.context.globalState.get<string | undefined>(LogLevelMonitor.logLevelChangedStorageKey);
+
+		if (lastChange) {
+			const date = new Date(lastChange);
+			if (date instanceof Date && !isNaN(date.valueOf())) {
+				return date;
+			}
+		}
+		return undefined;
+	}
+
+	private get doNotPrompt(): boolean {
+		return this.context.globalState.get<boolean | undefined>(LogLevelMonitor.doNotPromptLogLevelStorageKey) || false;
+	}
+
+	private shouldNotifyExtendedLogging(): boolean {
+		// TODO: Should the fallback for this be 0, or new Date()?
+		// 0 Means every user that has logging enabled currently will be notified.
+		// new Date() means that users will not be notified until they have changed their
+		// logging setting at least once.
```

Comment:
This is the last thing I was unsure about.

Reply:
Let's use zero to notify existing users who have logging enabled but don't have a value for `LogLevelMonitor.logLevelChangedStorageKey`